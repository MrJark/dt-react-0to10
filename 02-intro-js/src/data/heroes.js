
// exportación "normal"
const heroes = [
    {
        id: 1,
        name: 'Batman',
        owner: 'DC'
    },
    {
        id: 2,
        name: 'Spiderman',
        owner: 'Marvel'
    },
    {
        id: 3,
        name: 'Superman',
        owner: 'DC'
    },
    {
        id: 4,
        name: 'Flash',
        owner: 'DC'
    },
    {
        id: 5,
        name: 'Wolverine',
        owner: 'Marvel'
    },
];
// exportación por defecto:
// no se pone el nombre de una constante ni la palabra reservada. A la hora de exportarlo, al archivo que lo vayamos a exportar es cuando le damos el nombre a este 
// import patatasConJamon from './....... '; entonces este archivo donde lo estoy llamando, tendría el nombre de patatasConJamon
// export default [
//     {
//         id: 1,
//         name: 'Batman',
//         owner: 'DC'
//     },
//     {
//         id: 2,
//         name: 'Spiderman',
//         owner: 'Marvel'
//     },
//     {
//         id: 3,
//         name: 'Superman',
//         owner: 'DC'
//     },
//     {
//         id: 4,
//         name: 'Flash',
//         owner: 'DC'
//     },
//     {
//         id: 5,
//         name: 'Wolverine',
//         owner: 'Marvel'
//     },
// ];

export const owners = ['DC', 'Marver',];
export default heroes;

// Tmabién se pueden hacer exportaciones en grupo
// export {
//     heroes as default, // si quieres exportar algo por defecto y que esté en el grupo, tienes que hacerlo de esta manera pero ten en cuneta que al importarlo, tendrá que ir sin {}
//     owners
// }