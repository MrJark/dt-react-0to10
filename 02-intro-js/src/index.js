
// No es recomendable crear funciones de esta manera ya que son muy "dÃ©biles"
// function saludar(name) {
//     return `Hola, ${name} `;
// };

// Forma correcta
const saludar = function (name) {
    return `Hola, ${name} `;
};


// Funciones de flecha
const saludar2 =(name) => {
    return `Hola, ${name}`;
};

console.log(saludar2('Vegeta'));