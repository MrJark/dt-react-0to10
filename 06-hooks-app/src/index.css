body {
    padding: 12px;
    background: #222222;
    color: #F9F9F9;
}

button {
    margin-right: 4px;
}
h1,
h2,
h3,
h4 {
    color: #6AD2D3;
}

p {
    color: #F9F9F9;
}

.onColums {
    display: flex;
}

.btn-container {
    margin-left: 24px;
}

.names {
    margin-right: 16px;
}
img {
    border-radius: 12px;
}
.btn-toggle-modi {
    border-radius: 100%;
    width: 25px;
    height: 25px;
    align-self: center;
    
    border: none;
    background: #cacaca;
}
.btn-toggle-modi:hover {
    background: #8c8c8c;
    border-color: #222222;
    box-shadow: 2px 2px 2px #383838;

}