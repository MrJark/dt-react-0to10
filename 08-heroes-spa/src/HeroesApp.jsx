import { AuthProvider } from './auth';
import { AppRouter } from './router/AppRouter';

export const HeroesApp = () => {
    return (
        // el AuthProvider en esta ruta tan alta es para que coja todos los elementos
        <>
            <AuthProvider>
                <AppRouter/>
            </AuthProvider>
        </>
    )
}

