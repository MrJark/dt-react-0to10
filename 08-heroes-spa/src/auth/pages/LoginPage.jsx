import { useNavigate } from 'react-router-dom';


export const LoginPage = () => {

  const navigate = useNavigate(); // para poder navegar entre rutas más fácil sin tener que crearte tu custom hook

  const onLogin = () => {
    navigate('/', {
      replace: true, // esto es para remplazar el historial. Para que no pueda regresar a la ruta anterior
      
    });

  };

  return (
    <>
      <div className="container mt-5">
        <h1>Login</h1>
        <hr />
        <button 
          className="btn btn-secondary"
          onClick={ onLogin }
        >
        Login</button>
      </div>
    </>
  )
}
