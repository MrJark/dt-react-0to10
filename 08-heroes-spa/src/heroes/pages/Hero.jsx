import { Navigate, useParams } from 'react-router-dom';
import { getHeroById } from '../helpers';


export const Hero = () => {

    // const params = useParams(); // hook creado por react-router que sirve para obtener los parámetros
    // console.log(params);
    const { id } = useParams(); // hook creado por react-router que sirve para obtener los parámetros
    // console.log(id);
    

    const hero = getHeroById(id);

    if( !hero ) {
        return <Navigate to="/marvel"/>
    }

    return (
        <>
            <h1>{hero.superhero}</h1>
        
        </>
    )
}
