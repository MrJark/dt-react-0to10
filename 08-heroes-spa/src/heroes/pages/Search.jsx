import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks';

import { HeroCard } from '../components';
import { getHeroByName } from '../helpers';


export const Search = () => {

    // hooks personalizados de react router dom
    const navigate = useNavigate();
    const location = useLocation(); 

    // const query = queryString.parse( location.search ); 
    const { q = '' } = queryString.parse( location.search ); // para parsear se una manera mas sencilla
    
    const heroes = getHeroByName(q); // para coger el hero con el query que puso el usuario
   
    
    const { searchText, onInputChange } = useForm({ // hook creado anteriormente y que se puede utilizar para controlar el formulario
        searchText: q,
    });

    const onSearchSub = ( e ) => { // la e es el evento
        e.preventDefault();
        if( searchText.trim().length <= 1) return;

        // console.log({searchText});
        navigate(`?q=${ searchText }`);
    }

    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className="row animate__animated animate__fadeIn">
                <div className="col-5">
                    <h4>Searching</h4>
                    <hr />

                    <form onSubmit={ onSearchSub }>
                        <input 
                            type="text" 
                            placeholder="Search a hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ onInputChange }
                        />

                        <button 
                            className="btn btn-outline-secondary mt-2"
                        >
                        Search</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    <div className="alert alert-secondary">
                        Search a hero
                    </div>
                    <div className="alert alert-danger">
                        There's not a hero with <b>{ q }</b>
                    </div>
                    {
                        heroes.map( hero => (
                            <HeroCard key={ hero.id } { ...hero }/>
                        ))
                    }
                </div>    
            </div> 
            
        </>
    )
}
