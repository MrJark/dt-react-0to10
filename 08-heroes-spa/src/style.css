:root {
  --black: #020D0D;
  --white: #f9f9f9;
  --primary: #0C9A9D;
  --socondary: #ab79dd;
  --background-dark: #222222;
  --background-dark-light: #404141;

}

body {
  padding: 12px;
  font-family: 'Ubuntu', Verdana, Tahoma, sans-serif;
  background: var(--background-dark);
  color: var(--white);
  letter-spacing: 1px;
}

h1, h2, h3, h4 {
  color: var(--primary);
}

.navbar-bg-modified {
  background: var(--background-dark-light);
}

.navbar-span {
  color: var(--socondary);
}