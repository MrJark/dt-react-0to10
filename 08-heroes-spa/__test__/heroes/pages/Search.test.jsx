import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { Search } from '../../../src/heroes/pages';
import { AuthContext } from '../../../src/auth';


// para la Task 2 necesitas un mockup parcial de la ruta useNavegate ( esto si lo hab√≠a hecho bien )
const mockedUseNavegate = jest.fn();
jest.mock('react-router-dom', () => ({ // mock de la ruta que necesitas
    ...jest.requireActual('react-router-dom'), // spread de todas las rutas que usas en el test
    useNavigate: () => mockedUseNavegate // pero le dices que solo quieres esta y eso es una funci√≥n de ah√≠ que se cree la const mockedUseNavegate
}));

describe(' test on <Search/>', () => {

    const infoForm = {};

    test('should show default value with snapshot', () => {

        const { container } = render(
            <MemoryRouter>
                <Search/>
            </MemoryRouter>
        );
        expect( container ).toMatchSnapshot();
        // en esto y ver la app, la pag se search, sabes si est√° bien y si lo est√°, haces el snapshot
        // render(
        //     <MemoryRouter>
        //         <Search/>
        //     </MemoryRouter>
        // );
        // screen.debug(); 

    });

    test('should show Batman and the input in queryString', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Search/>
            </MemoryRouter>
        );
        // screen.debug()
        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman'); // mostrar el input en el query
        
        // Tarea: mostrar qie los divs no se muestran. Es decir, display: none. (üî∏ Casi lo consigo, el error a sido querer coger los dos divs a la vez, tenia que hacerlo de uno en uno pero la l√≥gica era la misma, pero con el getByLabelText y no con el All)
        // con esta validaci√≥n siguiente no es suficiente ya que el div que diga no se encontr√≥ y el de busca, deber√≠an tener el display en none y para hacerlo tienes que colocar un aria-label que los identifique y puedas 'cogerlos' con un getAllByLabelText
        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/dc-batman.jpg'); 
        // expect(screen.getAllByTestId('catch-display-none')).toContain('none')
        const div = screen.getByLabelText('catch-display-none-search');
        // console.log(div.style);
        expect( div.style.display ).toBe('none');
    });

    test('Task 1: should show the error div when it doesnt find a hero ‚úÖ', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Search/>
            </MemoryRouter>
        );
        const divError = screen.getByLabelText('catch-display-none-not-hero');
        // console.log(divError.style);
        expect( divError.style ).not.toContain('none'); // el mio
        expect( divError.style.display ).toBe(''); // el del profesor

    });

    test('Task 2: should the navigate call a new screen (superman card) ‚ùå ', () => {

        // render( // lo que estaba intentando y no estaba bien ü•≤
        //     <AuthContext.Provider value={'superman'}>
        //         <MemoryRouter initialEntries={['/search?q=superman']}>
        //             <Search/>
        //         </MemoryRouter>
        //     </AuthContext.Provider>
        // );

        const inputValue = 'superman'; // forma 2 de hacerlo y la cual me da una mayor flexibilidad y mejor control

        render(
            <MemoryRouter initialEntries={['/search']}>
                <Search/>
            </MemoryRouter>
        );

        const inputBtn = screen.getByRole('textbox');
        fireEvent.change( inputBtn, { 
            target: {
                name: 'searchText', 
                // value: 'superman' // forma 1 de hacerlo
                value: inputValue
            }
        }) // al pulsar el bot√≥n, tiene que cambiar (.change(los valores) )

        const formForm = screen.getByLabelText('to-test'); // para conseguir que la const apunte al formulario ya que aun no hay un rol que vaya al form
        fireEvent.submit( formForm);

        // expect( mockedUseNavegate ).toHaveBeenCalledWith('?q=superman'); // (Forma1 de hacerlo) es llamado con este query porque le he puesto que sea superman en el value
        expect( mockedUseNavegate ).toHaveBeenCalledWith(`?q=${inputValue}`); // Forma 2 de hacerlo es llamado con este query porque le he puesto que sea superman en el value

    })

});