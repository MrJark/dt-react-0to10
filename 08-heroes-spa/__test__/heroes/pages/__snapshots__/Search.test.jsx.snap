// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` test on <Search/> should show default value with snapshot 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row animate__animated animate__fadeIn"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="to-test"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-secondary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <div
        aria-label="catch-display-none-search"
        class="alert alert-secondary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="catch-display-none-not-hero"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There's not a hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
